### ISSUE TITLE
Uplift to .NET Framework 4.8

### ISSUE THREAD

**Issue by TheCakeIsNaOH** [chocolatey/choco#2738]
Since older .Net framework versions are going out of support, and because newer versions of the NuGet libraries require newer .Net versions (#508), Chocolatey CLI should be upgraded to be built on .Net 4.8

This is also a stepping stone to upgrading to dotnet 5+, #2436 

This will also allow upgrading the SimpleInjector version, which likely would be helpful for #2572 and for #2693

- [x] As recently linked, uplifting NUnit to v3.x
- [x] ~~Updating the initial install.ps1 script to check for and install .net 4.8 if required~~ Only adding check to `chocolateySetup.psm1`
- [x] Updating the Chocolatey CLI package install/upgrade process to validate that .net 4.8 is installed. https://github.com/chocolatey/choco/pull/2933
- [ ] Updating documentation on the docs site, and the main site for Chocolatey CLI requirements.
  - https://github.com/chocolatey/docs/issues/638
  - https://github.com/chocolatey/chocolatey.org/issues/228
- [ ] Validating that all currently supported operating systems work with .net 4.8. I think they all do, but I'm not sure if there is a windows server version or something that is not compatible. 
  - https://github.com/chocolatey/choco/issues/2992 
  - It bumps the requirement to 2008 R2 https://github.com/chocolatey/choco/issues/2903
- [x] Add a check into chocolatey.console that .net 4.8 is installed. https://github.com/chocolatey/choco/pull/2935
- [ ] Investigate if there are any other dependency libraries that can/should be uplifted or migrated to an alternative. I think this is something that does not need to happen before .net 4.8 version of choco comes out. https://github.com/chocolatey/choco/issues/2896
- [ ] Update TLS handling https://github.com/chocolatey/choco/issues/2902 https://github.com/chocolatey/choco/issues/2585

**Comment by TheCakeIsNaOH** [comment]
There are other things that should be done as a part of uplifting Chocolatey CLI to .net 4.8 that are outside the scope of this issue (which is uplifting the actual `choco.exe` build process).
- As recently linked, uplifting NUnit to v3.x
- Updating the initial `install.ps1` script to check for and install .net 4.8 if required
- Updating the Chocolatey CLI package upgrade process to validate that .net 4.8 is installed (is a reboot required after install in any cases?)
- Updating documentation on the docs site, and the main site for Chocolatey CLI requirements.
- Validating that all currently supported operating systems work with .net 4.8. I think they all do, but I'm not sure if there is a windows server version or something that is not compatible.
- Investigate if there are any other dependency libraries that can/should be uplifted or migrated to an alternative. I think this is something that does not need to happen before .net 4.8 version of choco comes out.
- Maybe something I am missing?

**Comment by gep13** [comment]
@TheCakeIsNaOH thanks for capturing all those things.  I am going to move them up as tasks within this issue so that they don't get over looked.

**Comment by gep13** [comment]
@TheCakeIsNaOH The core work of changing the target .NET Framework version is completed, and there are follow up issues created for the remaining work that was highlighted in this issue.  As such, I am going to go ahead and close this issue out.

Let me know if you think I have missed anything.

**Comment by slonopotamus** [comment]
This broke installation of Chocolatey inside 2019ltsc Docker image: https://github.com/adamrehn/ue4-docker/actions/runs/5136725520/jobs/9243844531?pr=322#step:5:119

**Comment by pauby** [comment]
Please see the [upgrade guide](https://docs.chocolatey.org/en-us/guides/upgrading-to-chocolatey-v2-v6#net-framework-4.8-required). I'd also suggest that you install a known working version rather than always installing the latest.

**Comment by slonopotamus** [comment]
I've read through upgrade guide (thanks for the link). The only relevant thing I see there is that one should reboot after installing .NET4.8. It isn't clear what "reboot" means in the context of Docker container.

Would you like me to create a separate issue with clear repro steps?

**Comment by pauby** [comment]
The solution for Docker would be to use a Docker image that has .NET 4.8 already installed.

> Would you like me to create a separate issue with clear repro steps?

This isn't a Chocolatey CLI issue but a Docker container one. You need to have .NET 4.8 installed in the container to use Chocolatey CLI 2.0.0. The alternative is to stick to Chocolatey CLI 1.4.0.

**Comment by slonopotamus** [comment]
> This isn't a Chocolatey CLI issue but a Docker container one.

I would agree with this if Chocolatey 2.0.0 installer just errored out with "you need .NET 4.8 as a prerequisite to use this thing, bye". But Chocolatey installer is trying to install .NET and fails to proceed after that. So _possibly_ there exists a solution involving tweaks to the installer script that would make it to Just Work.

Anyway, I'm going to follow your suggested workarounds for now, no problem.

**Comment by pauby** [comment]
> But Chocolatey installer is trying to install .NET and fails to proceed after that. So possibly there exists a solution involving tweaks to the installer script that would make it Just Work.

Chocolatey has installed .NET 4.8 on your Docker container, as it was not already installed. It then indicates that a reboot is required and doesn't reboot for you (to be safe).

![image](https://github.com/chocolatey/choco/assets/12760779/657368ac-127a-4bd7-8362-a0a65614283d)

A Docker container cannot be rebooted, so you need to use a container with .NET 4.8 installed already. We provide a [Docker container](https://hub.docker.com/r/chocolatey/choco) for you to use. 

If you create your own, then please understand the requirements of the software and, as I said, only use a known working version. Installing the latest for any software is going to always cause you problems somewhere along the road.
